File: javascript/HelloApp.test.js
const HelloMessage = require("./HelloMessage");
const HelloConsole = require("./HelloConsole");
const HelloApp = require("./HelloApp");

jest.mock('./HelloMessage');
jest.mock('./HelloConsole');

describe('HelloApp', () => {
  it("should print hello message", () => {
    const messageText = "Hello Test!";
    HelloMessage.mockImplementation(() => {
      return {
        text: messageText
      }
    });
    const message = new HelloMessage();

    const console = new HelloConsole();

    const app = new HelloApp(message, console);
    app.printHello();

    expect(console.print).toHaveBeenCalledTimes(1);
    expect(console.print).toHaveBeenCalledWith(messageText);
  })
});

File: go/HelloApp_test.go
package main

import "testing"

type HelloMessageMock struct {
    text string
}

func (message *HelloMessageMock) Text() string {
    return message.text
}

type HelloConsoleMock struct {
    Calls int
    Text  string
}

func (console *HelloConsoleMock) Print(text string) {
    console.Calls++
    console.Text = text
}

func TestShouldReturnPrintHelloMessage(t *testing.T) {
    messageText := "Hello Test!"
    message := HelloMessageMock{messageText}

    console := HelloConsoleMock{}

    app := HelloApp{&message, &console}
    app.PrintHello()

    if console.Calls != 1 {
        t.Fatalf("HelloConsole expected calls 1 but got %d", console.Calls)
    }

    if console.Text != messageText {
        t.Fatalf("HelloConsole expected text %s but got %s", messageText, console.Text)
    }
}
